In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
import matplotlib.pyplot as plt

In [5]:
import seaborn as sns

In [6]:
titanic=sns.load_dataset("titanic")  ## to load the data

In [10]:
titanic.head()   ##.head() :- for checking the data as a sample

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [12]:
titanic.info()   ## info to check how many rows and column are available

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [13]:
#As we can check,some are miising values in the data ,so lets checkthe exact number.

In [15]:
titanic.isnull().sum()  ## isnull().sum() :-- to check themissing values (exact number)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

as we check ,
1) age-177 are missing values,should be imported basd on pclass or sex(because it is a numerical values replace with mean or median)

2) embarked town - 2 missing values can be filled with mode (because it is categorical values thats y we replace with mode)

3) deck-massive missing values (value in nun form,unavailble values in the dataset.better to drop this column.


In [16]:
## age filled with median based on pclass or sex

1).Transform :-- gives the value in series form (gives the result in same shape as the original data.

2) .apply :- make a data frame.

3).fillna :- if value is missing then it will fill the value.

4). drop :- it will drop or delete the value permantly.





In [20]:
titanic['age']=titanic.groupby(['pclass','sex'])['age'].transform(lambda x : x .fillna(x.median()))


1.   transform(lambda x: x.fillna(x.median())): For each group, it applies a function to fill in missing age values (NaNs). It fills these missing values with the median age of that specific group (same "pclass" and "sex").

The median is the middle value in a sorted list of numbers, so it's a good way to fill in missing data without being affected by extreme values (outliers).





In [22]:
## Replace embark with mode :- mode:- most frequent values

In [31]:
titanic['embarked'].mode()[0]  ## [0] is used to get the first value in that list, which is the most common value.


'S'

In [34]:
titanic['embarked'].fillna(titanic['embarked'].mode()[0],inplace=True)

In [36]:
titanic["embarked"].isnull().sum()  ## for checking the missing values

0

In [37]:
##deck: drop the column (delete)

In [40]:
titanic.drop(columns=['deck'],inplace=True)

In [41]:
titanic.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [42]:
##drop the unwanted column (delete)

In [44]:
titanic.drop(columns=['alive','who','embark_town','adult_male'],inplace=True)

In [45]:
titanic.isnull().sum()  ##to check all the missing values

survived    0
pclass      0
sex         0
age         0
sibsp       0
parch       0
fare        0
embarked    0
class       0
alone       0
dtype: int64

In [46]:
# to check rows and column

In [47]:
titanic.shape

(891, 10)

After dropping the unwanted column,there are 891 rows ,describe the passanger details and 10 column which shows the passanger features.

In [55]:
titanic.describe(include='all')  # for statstical summary

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  891.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.112424    0.523008    0.381594   
std       0.486592    0.836071   NaN   13.304424    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   21.500000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   26.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   36.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class alone  
count   891.000000      891    891   891  
unique         NaN        3      3     2  
top            NaN        S  Third  True  
freq           NaN      646    491   537  
mean     32.204208      NaN    NaN   NaN  
std      49.693429      NaN    NaN   NaN  
min       0.000000      NaN    NaN   NaN  
25%       7.910400      NaN    NaN   NaN  
50%      14.454200      NaN    NaN   NaN  
75%      31.000000      NaN    NaN   NaN  
max     512.329200      NaN    NaN   NaN

In [51]:
#check the data and write the insight

-- servival rate only 38% ,remaining 98% did not survive.

-- most of passanger were in third class.

-- most of passanger were young (median=26 year) (using 98-95-99.7% rule ,29+13(1st std)+13(2nd std),+13(3rd std)= 68 years (max age) --but shows 80years old also

-- It must be outlier in age becaue it is showing 80 years old also

-- fare=we will take  50% = 15+50(1st std)+50(2nd std)+50(3rd std)=165 ,and max fare was 512 (so outlier) ----------------------------------------------------
median fair 14.45 but max was 512.33.it also may contain the outliers(empirial rule)

embarked-most of passanger from "S"

Alone:- most of passanger travel alone.

class:- most passanger class were in third.


In [61]:
## to check duplicates with count

In [66]:
titanic.duplicate().sum()  ## it is alredy delted thats y showing like this but this is actual code before deleting

AttributeError: 'DataFrame' object has no attribute 'duplicate'

In [67]:
titanic.drop_duplicates(inplace=True)

In [68]:
titanic.duplicated().sum()   ## remove all the duplicate values.

0